tests/PHPStan/Analyser/data/bug-7663.php

```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Bug7663;
+
+declare(strict_types=1);
+
+use function PHPStan\Testing\assertType;
+
+class HelloWorld
+{
+	/**
+	 * @param 'de_DE'|'en_US' $language
+	 */
+	public function sayHello($language): void
+	{
+		assertType("'de'|'en'", substr($language, 0, 2));
+		assertType("'de_DE'|'en_US'", substr($language, 0, 10));
+
+		assertType("'DE'|'US'", substr($language, 3));
+		assertType("'_DE'|'_US'", substr($language, -3));
+		assertType("'_'", substr($language, -3, 1));
```

Comment:
it might be worth adding an `assertType` for the `false` return case when asking for invalid length offsets?

also `substr` in php7 and php8 return different values when given invalid `$length` or `$offsets`; 7 returns `false` whereas 8 returns `''`:

https://www.php.net/manual/en/function.substr.php#refsect1-function.substr-changelog

Reply:
agreed, added a few cases.